core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
By the way, this actually only saves as an RPC if we're caching blocks. If we're not caching anything, it will still be 1 RPC every time an executor is idle. If the latter turns out to be not a significant load on the driver after all, then maybe it makes sense to simplify the code by getting rid of this map.

I can't tell whether it's totally worth it off the top of my head. It would be good to test this with many executors (order hundreds) to investigate whether it actually saves you much.